### DISCUSSION TITLE
Disk space usage without fragmentation

### DISCUSSION THREAD

**Discussion by andmat900** [etcd-io/etcd#18305] [Q&A] [answered]
I have a database with frequent writes/deletes of keys. Auto compaction is set to one revision. After a key is deleted, will it's space be used to write the next key before allocating new disk space?

Without defragmentation, is there a risk that etcd will gradually eat up disk space even if the actual data size is much smaller than the partition size?

**Comment by ahrtr** [comment, marked as answer]
> will it's space be used to write the next key before allocating new disk space?

Yes, the freed pages can be reused. It will allocate more disk space if the existing free pages can't meet the allocation request. But please also note that bbolt (storage engine used by etcd) requires continuous pages each time when it tries to allocates pages. 



> Without defragmentation, is there a risk that etcd will gradually eat up disk space even if the actual data size is much smaller than the partition size?

It's possible, but low.
